File: leetcode/recursion + memo/97. Interleaving String/js/main.js
// Time O(N*M)
// Space O(N*M)
const isInterleave = (s1, s2, s3) => {
  if (s1.length + s2.length < s3.length) {
    return false;
  }

  if (s1.length == 0 || s2.length == 0) {
    return s1 == s3 || s2 == s3;
  }

  let dp = Array(s1.length)
    .fill(0)
    .map(() => Array(s2.length).fill(null));

  return dfs(0, 0, 0);

  function dfs(sIndex1, sIndex2, sIndex3) {
    if (sIndex1 >= s1.length) {
      return s2.substring(sIndex2) == s3.substring(sIndex3);
    }

    if (sIndex2 >= s2.length) {
      return s1.substring(sIndex1) == s3.substring(sIndex3);
    }

    if (dp[sIndex1][sIndex2] != null) {
      return dp[sIndex1][sIndex2];
    }

    if (
      (s1[sIndex1] == s3[sIndex3] && dfs(sIndex1 + 1, sIndex2, sIndex3 + 1)) ||
      (s2[sIndex2] == s3[sIndex3] && dfs(sIndex1, sIndex2 + 1, sIndex3 + 1))
    ) {
      dp[sIndex1][sIndex2] = true;

      return dp[sIndex1][sIndex2];
    }

    dp[sIndex1][sIndex2] = false;

    return dp[sIndex1][sIndex2];
  }
};

File: leetcode/recursion + memo/97. Interleaving String/py/main.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        if len(s1) + len(s2) != len(s3):
            return False

        if len(s1) == 0 or len(s2) == 0:
            return s1 == s3 or s2 == s3

        dp = [[None] * len(s2) for _ in range(len(s1))]
        
        def dfs(s_index1, s_index2, s_index3):
            if s_index1 >= len(s1):
                return s2[s_index2:] == s3[s_index3:]

            if s_index2 >= len(s2):
                return s1[s_index1:] == s3[s_index3:]
            
            if dp[s_index1][s_index2] != None:
                return dp[s_index1][s_index2]

            if (s1[s_index1] == s3[s_index3] and dfs(s_index1 + 1, s_index2, s_index3 + 1)) or (
                s2[s_index2] == s3[s_index3] and dfs(s_index1, s_index2 + 1, s_index3 + 1)
            ):
                dp[s_index1][s_index2] = True
                return True

            dp[s_index1][s_index2] = False
            return False

        return dfs(0, 0, 0)
